tests/integration/ha_tests/test_self_healing.py

```diff
@@ -47,6 +47,63 @@ async def test_build_and_deploy(ops_test: OpsTest) -> None:
         await ops_test.model.wait_for_idle(status="active", timeout=1000)
 
 
+@pytest.mark.ha_self_healing_tests
+@pytest.mark.parametrize("process", [POSTGRESQL_PROCESS])
+async def test_freeze_db_process(
+    ops_test: OpsTest, process: str, continuous_writes, master_start_timeout
+) -> None:
+    # Locate primary unit.
+    app = await app_name(ops_test)
+    primary_name = await get_primary(ops_test, app)
+
+    # Freeze the database process.
+    await send_signal_to_process(ops_test, primary_name, process, "SIGSTOP")
+
+    async with ops_test.fast_forward():
+        # Verify new writes are continuing by counting the number of writes before and after a
+        # 2 minutes wait (a db process freeze takes more time to trigger a fail-over).
+        writes = await count_writes(ops_test, primary_name)
+        for attempt in Retrying(stop=stop_after_delay(60 * 2), wait=wait_fixed(3)):
+            with attempt:
+                more_writes = await count_writes(ops_test, primary_name)
+                assert more_writes > writes, "writes not continuing to DB"
+
+        # Un-freeze the old primary.
+        await send_signal_to_process(ops_test, primary_name, process, "SIGCONT")
+
+        # Verify that the database service got restarted and is ready in the old primary.
+        assert await postgresql_ready(ops_test, primary_name)
+
+    # Verify that a new primary gets elected (ie old primary is secondary).
+    new_primary_name = await get_primary(ops_test, app, down_unit=primary_name)
+    assert new_primary_name != primary_name
```

Comment:
move this to before "Un-freeze old primary"

Reply:
Thanks! I'll do it.

Reply:
Updated on https://github.com/canonical/postgresql-k8s-operator/pull/49/commits/c07b3fc338114ff75299f9a4fc2d0efaffd9c70f.